### PR TITLE
53 LDAP query bundle metadata

### PR DESCRIPTION
**Pull request by kevinleeMW**
Changes to allow LDAP query on bundle properties

Requirements

R1. Use LDAP queries against bundle/framework meta-data.
R2. Allow customization of LDAP search strategies against meta-data without impacting existing client code. 
R3. Make it easy and intuitive for C++ developers to query meta-data.
R4. LDAP query performance for full string match

### CODE REVIEW

**Comment by jeffdiclemente** [comment]
What are the requirements for this feature? 
Knowing the requirements allows code reviewers to give more meaningful feedback.

File: core/src/util/usProperties.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    
+    keys.reserve(p.size());
+    values.reserve(p.size());
+    
+    for (std::map<std::string, Any>::const_iterator iter = p.begin();
+         iter != p.end(); ++iter)
+    {
+        if (Find(iter->first) > -1)
+        {
+            std::string msg = "BundleProperties object contains case variants of the key: ";
```
Comment (jeffdiclemente) [inline core/src/util/usProperties.cpp]: This message is inconsistent with the one on line 67.

File: core/test/usLDAPQueryTest.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    int actualMatchCount = 0;
+    int expectedMatchCount = 1;
+
+    for (std::shared_ptr<Bundle> bundle : bundles)
+    {
+        
+        if (bundle->GetBundleId() != 0)
+        {
+            bundle->Start();
+            //std::cout << bundle->GetProperty("bundle.name").ToString() << std::endl;
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: I'm assuming this no longer serves a purpose? If so, I'd suggest removing this commented code.

File: core/test/usTestDriverActivator.cpp [hunk 3]
```diff
@@ -57,6 +57,7 @@ US_IMPORT_BUNDLE(CppMicroServices)
 US_IMPORT_BUNDLE(TestBundleA)
 US_IMPORT_BUNDLE(TestBundleA2)
 US_IMPORT_BUNDLE(TestBundleB)
+US_IMPORT_BUNDLE(TestBundleLQ_1)
```
Comment (jeffdiclemente) [inline core/test/usTestDriverActivator.cpp]: Shouldn't all the other "TestBundleLQ_" bundles be listed here as statically linked bundles as well?

File: core/src/util/usLDAPFilter.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+bool LDAPFilter::Match(const ServiceReferenceBase& reference) const
+{
+  return d->ldapExpr.Evaluate(reference.d->GetProperties(), true);
+}
+    
+bool LDAPFilter::Match(const Bundle& bundle) const
+{
+  return d->ldapExpr.Evaluate(Properties(bundle.d->bundleManifest.GetProperties()), true);
```
Comment (jeffdiclemente) [inline core/src/util/usLDAPFilter.cpp]: Why not call ```bundle.d->bundleManifest.GetProperties()``` instead? 
Doing this also eliminates adding LDAPFilter as a friend class to ```Bundle```.

File: core/src/util/usLDAPExpr.cpp [hunk 5]
```diff
@@ -0,0 +1,852 @@
+/*=============================================================================
```
Comment (jeffdiclemente) [inline core/src/util/usLDAPExpr.cpp]: as we discussed offline, see if you can make git recognize this as a move of a file instead of a whole new file. If that is impossible, please summarize what areas have changed for code review. Thanks.

File: core/src/util/usLDAPExpr_p.h [hunk 6]
```diff
@@ -0,0 +1,180 @@
+/*=============================================================================
```
Comment (jeffdiclemente) [inline core/src/util/usLDAPExpr_p.h]: as we discussed offline, see if you can make git recognize this as a move of a file instead of a whole new file. If that is impossible, please summarize what areas have changed for code review. Thanks.

File: core/src/util/usLDAPFilter.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
+/*=============================================================================
```
Comment (jeffdiclemente) [inline core/src/util/usLDAPFilter.cpp]: as we discussed offline, see if you can make git recognize this as a move of a file instead of a whole new file. If that is impossible, please summarize what areas have changed for code review. Thanks.

File: core/src/util/usProperties_p.h [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  int FindCaseSensitive(const std::string& key) const;
+
+  const std::vector<std::string>& Keys() const;
+
+private:
+
+  std::vector<std::string> keys;
+  std::vector<Any> values;
+
+  static Any emptyAny;
```
Comment (jeffdiclemente) [inline core/src/util/usProperties_p.h]: ```static Any emptyAny``` should be ```const``` as it doesn't appear to change throughout its lifetime.

File: core/src/util/usProperties_p.h [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+=============================================================================*/
+
+#ifndef USPROPERTIES_P_H
+#define USPROPERTIES_P_H
+
+#include "usServiceProperties.h"
```
Comment (jeffdiclemente) [inline core/src/util/usProperties_p.h]: headers should be self sufficient (i.e. they include everything they need for their declaration). See #66 for reasons why.

You should add ```<vector>```, ```<map>```, ```<string>``` and ```Any``` to satisfy the declaration.

File: core/src/util/usProperties.cpp [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  {
+    if (Find(iter->first) > -1)
+    {
+      std::string msg = "ServiceProperties object contains case variants of the key: ";
+      msg += iter->first;
+      throw std::runtime_error(msg.c_str());
+    }
+    keys.push_back(iter->first);
+    values.push_back(iter->second);
+  }
```
Comment (jeffdiclemente) [inline core/src/util/usProperties.cpp]: There looks to be common code shared between the two constructors. I'd suggest refactoring to hoist the common code into one place. This makes maintenance a little easier.

File: core/src/util/usProperties.cpp [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+const Any& Properties::Value(const std::string& key) const
+{
+  int i = Find(key);
+  if (i < 0) return emptyAny;
+  return values[i];
+}
+
+const Any& Properties::Value(int index) const
+{
+  if (index < 0 || static_cast<std::size_t>(index) >= values.size())
```
Comment (jeffdiclemente) [inline core/src/util/usProperties.cpp]: even for one line if/else statements, always using braces. This cuts down on inadvertently introducing failures.

File: core/test/usLDAPQueryTest.cpp [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+#include <usBundle.h>
+#include <usLDAPFilter.h>
+#include <usLog.h>
+
+#include "usTestUtils.h"
+#include "usTestingMacros.h"
+
+using namespace us;
+
+void TestLDAPQueryBundleMetaData(BundleContext* frameworkCtx)
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: If usTestBundleLQ_2 through usTestBundleLQ_5 do not export services and only contain meta-data, you shouldn't need to create dummy .cpp files for them. You should be able to generate "meta-data only" bundles. This will cut down on the number of files to maintain.

File: core/test/usLDAPQueryTest.cpp [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+    US_TEST_CONDITION(actualMatchCount == expectedMatchCount, test_description);
+}
+
+int usLDAPQueryTest(int /*argc*/, char* /*argv*/[])
+{
+    US_TEST_BEGIN("LDAPQueryTest");
+    
+    FrameworkFactory factory;
+    
+    std::shared_ptr<Framework> framework = factory.NewFramework(std::map<std::string, std::string>());
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: this can be simplified by removing ```std::map<std::string, std::string>()``` since no configuration is required.

File: core/test/usLDAPQueryTest.cpp [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    std::vector<std::shared_ptr<Bundle>> bundles = frameworkCtx->GetBundles();
+    LDAPFilter ldap( "(bundle.testproperty=YES)" );
+    
+    int actualMatchCount = 0;
+    int expectedMatchCount = 0;
+    
+    std::string bundleName;
+    Any bundleTestProperty;
+    
+    for (std::shared_ptr<Bundle> bundle : bundles)
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: I'd suggest using a ```const auto&``` for the range declaration to leverage type deduction and maintain const correctness. The ```bundle``` object will not be modified in the loop and ```LDAPFilter::Match(const Bundle& )``` expects a ```const&```.

File: core/test/usLDAPQueryTest.cpp [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+{
+    std::vector<std::shared_ptr<Bundle>> bundles = frameworkCtx->GetBundles();
+    LDAPFilter ldap( "(service.testproperty=YES)" );
+    
+    BundleContext* thisBundleCtx;
+    
+    int actualMatchCount = 0;
+    int expectedMatchCount = 1;
+
+    for (std::shared_ptr<Bundle> bundle : bundles)
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: This loop appears to be unnecessary. Correct me if I'm wrong; the point of this test looks to be to test that query service properties work. 
If that's that case, its inefficient to iterate over all the bundles which don't export a service (unless there is a requirement I'm missing). The same result can be reached by using ```BundleContext::GetBundle(const std::string& name)```, passing in the name of the bundle exporting the service.

File: core/test/usLDAPQueryTest.cpp [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+        
+            if (sr && ldap.Match(sr))
+            {
+                actualMatchCount++;
+            }
+            bundle->Stop();
+        }
+    }
+    
+    std::string test_description = "Test LDAP query of service properties (expected # of matches: " + std::to_string(expectedMatchCount) + "  actual # of matches: " + std::to_string(actualMatchCount)+ ")";
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: nit-pick: spacing of the ```+``` in ```std::to_string(actualMatchCount)+ ")";```

File: core/test/usLDAPQueryTest.cpp [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    InstallTestBundle(frameworkCtx, "TestBundleLQ_4");
+    InstallTestBundle(frameworkCtx, "TestBundleLQ_5");
+    
+    TestLDAPQueryBundleMetaData(frameworkCtx);
+    TestLDAPQueryServiceProperties(frameworkCtx);
+    
+    framework->Stop();
+    
+    US_TEST_END()
+}
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: nit-pick: no newline at end of the file.

File: core/include/usBundle.h [hunk 18]
```diff
@@ -398,6 +409,7 @@ class US_Core_EXPORT Bundle : public std::enable_shared_from_this<Bundle>
   friend class BundleRegistry;
   friend class ServiceReferencePrivate;
   friend class Framework;
+  friend class LDAPFilter;
```
Comment (jeffdiclemente) [inline core/include/usBundle.h]: can be removed if ```LDAPFilter::Match(...)``` uses ```Bundle::GetProperties()```

File: core/test/usLDAPQueryTest.cpp [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+    std::shared_ptr<Framework> framework = factory.NewFramework(std::map<std::string, std::string>());
+    framework->Start();
+    
+    BundleContext* frameworkCtx = framework->GetBundleContext();
+    
+    InstallTestBundle(frameworkCtx, "TestBundleLQ_1");
+    InstallTestBundle(frameworkCtx, "TestBundleLQ_2");
+    InstallTestBundle(frameworkCtx, "TestBundleLQ_3");
+    InstallTestBundle(frameworkCtx, "TestBundleLQ_4");
+    InstallTestBundle(frameworkCtx, "TestBundleLQ_5");
```
Comment (karthikreddy09) [inline core/test/usLDAPQueryTest.cpp]: Why 5 bundles ? Consider adding the absolute minimum number of bundles required to satisfy the test cases

File: core/include/usBundle.h [hunk 20]
```diff
@@ -244,6 +245,16 @@ class US_Core_EXPORT Bundle : public std::enable_shared_from_this<Bundle>
...
 
   /**
+   * Returns this bundle's Manifest properties as key/value pairs.
+   *
+   * @return A map containing this bundle's Manifest properties as 
+   *         key/value pairs.
+   *
+   * @sa \ref MicroServices_BundleProperties
+   */
+  std::map<std::string, Any> GetProperties() const;
```
Comment (karthikreddy09) [inline core/include/usBundle.h]: Any reason why we use an ordered map here and an unordered map for service properties ?

File: core/test/usLDAPQueryTest.cpp [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+    
+    BundleContext* thisBundleCtx;
+    
+    int actualMatchCount = 0;
+    int expectedMatchCount = 1;
+
+    for (std::shared_ptr<Bundle> bundle : bundles)
+    {
+        
+        if (bundle->GetBundleId() != 0)
```
Comment (karthikreddy09) [inline core/test/usLDAPQueryTest.cpp]: spacing

File: core/src/CMakeLists.txt [hunk 22]
```diff
@@ -65,13 +61,19 @@ set(_srcs
...
 set(_private_headers
   util/usFrameworkPrivate.h
+  util/usLDAPExpr_p.h
+  util/usProperties_p.h
   util/usThreads_p.h
   util/usUtils_p.h
   util/usWaitCondition_p.h
 
   service/usServiceHooks_p.h
+  service/usServiceListenerEntry_p.h
```
Comment (karthikreddy09) [inline core/src/CMakeLists.txt]: Is this change related to LDAP issue ? if not i'd suggest creating another branch for this issue. I believe this falls under issue#67

File: core/src/util/usLDAPExpr.cpp (hunk 5 above)
Comment (kevinleeMW) [inline core/src/util/usLDAPExpr.cpp]: I searched online, and found that it seems not possible with github UI. 
Summary of changes:
- use usProperties class instead of usServicePropertiesImpl class.

File: core/src/util/usLDAPExpr_p.h (hunk 6 above)
Comment (kevinleeMW) [inline core/src/util/usLDAPExpr_p.h]: I searched online, and found that it seems not possible with github UI. 
Summary of changes:
- use usProperties class instead of usServicePropertiesImpl class.

File: core/src/util/usLDAPFilter.cpp (hunk 7 above)
Comment (kevinleeMW) [inline core/src/util/usLDAPFilter.cpp]: I searched online, and found that it seems not possible with github UI. 
Summary of changes:
- use usProperties class instead of usServicePropertiesImpl class.
- Added "Match(const Bundle& bundle)" method to enable LDAP query on bundle properties

File: core/src/util/usLDAPExpr.cpp [hunk 23]
```diff
@@ -0,0 +1,852 @@
...
+{
+  return LDAPExpr(filter).Evaluate(pd, false);
+}
+
+bool LDAPExpr::Evaluate( const Properties& p, bool matchCase ) const
+{
+  if ((d->m_operator & SIMPLE) != 0)
+  {
+    // try case sensitive match first
+    int index = p.FindCaseSensitive(d->m_attrName);
```
Comment (karthikreddy09) [inline core/src/util/usLDAPExpr.cpp]: This is wierd ... why do attempt a casesentive match at all if matchCase is false.

File: core/src/util/usLDAPFilter.cpp (hunk 4 above)
Comment (karthikreddy09) [inline core/src/util/usLDAPFilter.cpp]: +1. Bundle should have a GetProperties public method instead

File: core/include/usBundle.h (hunk 20 above)
Comment (kevinleeMW) [inline core/include/usBundle.h]: I will address this issue in the second iteration of this task (incl. optimization, etc.)

File: core/src/util/usLDAPExpr.cpp (hunk 5 above)
Comment (saschazelzer) [inline core/src/util/usLDAPExpr.cpp]: Some files like `usLDAPExpr.cpp` are still present in their original location (core/src/service). I guess this is the reason why the changes are not schown as a move. Could you remove the files, squash the commits and push-force the branch again?

File: core/src/util/usProperties.cpp (hunk 11 above)
Comment (saschazelzer) [inline core/src/util/usProperties.cpp]: This is my old code, where I used one-line if statements without braces. Feel free to to add them though.

File: core/include/usBundle.h (hunk 20 above)
Comment (saschazelzer) [inline core/include/usBundle.h]: We could probably rename the `ServiceProperties` typedef (a std::unordered_map) to `Dictionary` and use this consistently.

File: core/src/util/usLDAPExpr.cpp (hunk 23 above)
Comment (saschazelzer) [inline core/src/util/usLDAPExpr.cpp]: This was optimized for the general case were it is assumed that people use filters on entities that contain the attribute names as property keys with the same case and the filter compares property values most of the time. Case mismatches are assumed to be typos or "laziness".

File: core/src/CMakeLists.txt (hunk 22 above)
Comment (kevinleeMW) [inline core/src/CMakeLists.txt]: These changes were made, because some of the files were moved; and made additional changes as I realized that some of the files were missing from this list.

File: core/test/usLDAPQueryTest.cpp [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+{
+  US_TEST_BEGIN("LDAPQueryTest");
+    
+  FrameworkFactory factory;
+  std::shared_ptr<Framework> framework = factory.NewFramework();
+  framework->Start();
+    
+  BundleContext* frameworkCtx = framework->GetBundleContext();
+    
+  InstallTestBundle(frameworkCtx, "TestBundleLQ");
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: ```InstallTestBundle``` returns the ```std::shared_ptr<Bundle>``` it installed. This makes the call to ```frameworkCtx->GetBundle``` on the next line superfluous.

File: core/test/usLDAPQueryTest.cpp [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+
+  bundle->Stop();
+}
+
+int usLDAPQueryTest(int /*argc*/, char* /*argv*/[])
+{
+  US_TEST_BEGIN("LDAPQueryTest");
+    
+  FrameworkFactory factory;
+  std::shared_ptr<Framework> framework = factory.NewFramework();
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: prefer using ```auto```

File: core/test/usLDAPQueryTest.cpp [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+  framework->Start();
+    
+  BundleContext* frameworkCtx = framework->GetBundleContext();
+    
+  InstallTestBundle(frameworkCtx, "TestBundleLQ");
+  const auto& bundle = frameworkCtx->GetBundle("TestBundleLQ");
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: why ```const auto&```?

File: core/test/usLDAPQueryTest.cpp [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+
+int usLDAPQueryTest(int /*argc*/, char* /*argv*/[])
+{
+  US_TEST_BEGIN("LDAPQueryTest");
+    
+  FrameworkFactory factory;
+  std::shared_ptr<Framework> framework = factory.NewFramework();
+  framework->Start();
+    
+  BundleContext* frameworkCtx = framework->GetBundleContext();
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: You could remove the variable and place ```framework->GetBundleContext()``` inside ```InstallTestBundle``` (assuming ```frameworkCtx->GetBundle``` is removed).

File: core/test/usLDAPQueryTest.cpp [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+  // Exact string match of both key and value
+  US_TEST_CONDITION(ldapMatchCase.Match(sr), " Evaluating LDAP expr: " + ldapMatchCase.ToString());
+
+  // Testing case-insensitive key (should still pass)
+  US_TEST_CONDITION(ldapKeyMismatchCase.Match(sr), " Evaluating LDAP expr: " + ldapKeyMismatchCase.ToString());
+
+  // Testing case-insensitive value (should fail)
+  US_TEST_CONDITION(!ldapValueMismatchCase.Match(sr), " Evaluating LDAP expr: " + ldapValueMismatchCase.ToString());
+
+  bundle->Stop();
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: What's the intended behavior of ```LDAPFilter::Match(ServiceReference)``` once the bundle has stopped (and the service reference is no longer valid)?
Does this behavior need to be tested?

File: core/test/usLDAPQueryTest.cpp [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+
+#include "usTestUtils.h"
+#include "usTestingMacros.h"
+#include "usProperties_p.h"
+#include "usServiceReferenceBasePrivate.h"
+
+
+using namespace us;
+
+void TestLDAPFilterMatchBundle(const std::shared_ptr<Bundle> bundle)
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: use ```const std::shared_ptr<Bundle>& bundle``` instead. That way there is no reference counting (which is unnecessary in this case and expensive) at function entry and exit.

File: core/test/usLDAPQueryTest.cpp [hunk 30]
```diff
@@ -0,0 +1,102 @@
...
+  US_TEST_CONDITION(ldapMatchCase.Match(*bundle), " Evaluating LDAP expr: " + ldapMatchCase.ToString());
+
+  // Testing case-insensitive key (should still pass)
+  US_TEST_CONDITION(ldapKeyMismatchCase.Match(*bundle), " Evaluating LDAP expr: " + ldapKeyMismatchCase.ToString());
+
+  // Testing case-insensitive value (should fail)
+  US_TEST_CONDITION(!ldapValueMismatchCase.Match(*bundle), " Evaluating LDAP expr: " + ldapValueMismatchCase.ToString());
+}
+
+void TestLDAPFilterMatchServiceReferenceBase(std::shared_ptr<Bundle> bundle)
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: use ```const std::shared_ptr<Bundle>& bundle``` instead. That way there is no reference counting (which is unnecessary in this case and expensive) at function entry and exit.

File: core/test/usLDAPQueryTest.cpp [hunk 31]
```diff
@@ -0,0 +1,102 @@
...
+ limitations under the License.
+ 
+ =============================================================================*/
+
+#include <usFrameworkFactory.h>
+#include <usFramework.h>
+
+#include <usBundle.h>
+#include <usLDAPFilter.h>
+#include <usLog.h>
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: is this needed? I don't see any invocations of the logging macros.

File: core/test/usLDAPQueryTest.cpp [hunk 32]
```diff
@@ -0,0 +1,102 @@
...
+
+#include "usTestUtils.h"
+#include "usTestingMacros.h"
+#include "usProperties_p.h"
+#include "usServiceReferenceBasePrivate.h"
+
```
Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: I don't believe you want to ```#include``` private headers here (```usProperties_p.h``` and ```usServiceReferenceBasePrivate.h```). The public headers should be sufficient.

File: core/test/usLDAPQueryTest.cpp (hunk 28 above)
Comment (kevinleeMW) [inline core/test/usLDAPQueryTest.cpp]: Quoting OSGi spec (Section 5.2.1):
"A ````ServiceReference```` object is valid only as long as the service is registered. However, its properties must remain available as long as the ````ServiceReference```` object exists."

From this, I think that ```LDAPFilter::Match(ServiceReference)```` should behave the same even after the bundle has stopped. 

Perhaps, an additional test condition can be added after the bundle stops to make sure that querying service properties via a ````ServiceReferenceBase```` object still works.

What do you guys think?

Comment (jeffdiclemente) [inline core/test/usLDAPQueryTest.cpp]: I think it sounds pretty clear that service properties can be queried after the service they reference has been unregistered (e.g. on bundle stop).

What happens if you duplicate the four ```US_TEST_CONDITION``` macros after the ```bundle->Stop();``` and re-run the test?

Comment (kevinleeMW) [inline core/test/usLDAPQueryTest.cpp]: The first of the four ````US_TEST_CONDITION```` macros will fail, because it checks to make sure that the ServiceReference object is valid, and it no longer is after the ````bundle->Stop();````. 

Including the second test macro (that does the querying and expects a match) after the ````bundle->Stop();```` seems sufficient, as it will be testing that the service properties are still available (for LDAP queries).

**Comment by kevinleeMW** [comment]
I went through and addressed all of the feedback from Jeff, Abhinay and Sascha. Some of the suggestions will be tracked in Issue #85 as a follow-up task.

Is there any other flags from the pull request, @jeffdiclemente @saschazelzer @karthikreddy09 ?

**Comment by jeffdiclemente** [comment]
I have no additional flags.
Thanks for creating #85 to further optimize.

**Comment by saschazelzer** [comment]
Looks good to me, +1 for merging.

**Comment by karthikreddy09** [comment]
Ship It !
